src/main/java/com/example/demo/service/PostsService.java

```diff
@@ -88,4 +95,83 @@ public Post unLike(Long id) {
         post.setLikeCount(max(0, post.getLikeCount() - 1));
         return postsRepository.save(post);
     }
+
+    // ------------------------------------------------------------------------------------------------------------------------------
+    // Function to check report Spam
+    public String spam(String description) {
+        Map<String, Integer> mp = new TreeMap<>();
+        int count = 0;
+        // Convert to array of words
+        String arr[] = description.split("[ .,-;*]+");
+
+        // Count the frequency of each word
+        for (int i = 0; i <= arr.length - 1; i++) {
+            if (mp.containsKey(arr[i])) {
+                mp.put(arr[i], mp.get(arr[i]) + 1);
+            } else {
+                mp.put(arr[i], 1);
+            }
+        }
+
+        // Count the number of word which is duplicated
+        for (Map.Entry<String, Integer> entry : mp.entrySet()) {
+            String key = entry.getKey();
+            Integer value = entry.getValue();
+            if (value != 1) count++;
+        }
+
+        if (count >= 2) return "NOT OK";
+
+        return "OK";
+    }
+
+    // Function to check report Impolite
+    public String impolite(String description) {
+        String[] keywords = appProperties.getKeywords();
+        Set<String> hashSet = new HashSet<String>();
+        // Convert to array of words
+        String[] arr = description.split("[ .,-;*]+");
+
+        // Convert to HASH SET of the word
+        Collections.addAll(hashSet, keywords);
```

Comment:
Bước này không cần thiết, có thể làm đơn giản hơn là for loop từng phần tử, nếu phát hiện ra thì return NOT OK luôn

Reply:
Do em lưu _keywords_ ở file _properties_ ... Khi lấy dùng thì nó được được lưu ở _array_ _keyword_
Khi for loop từng phần từ ở _arr_ , để phát hiện arr[i] có trong _keyword_ không em dùng _set_ để O(1) còn nếu dùng array thì mất O(n).

=> Em sẽ thử xem khi lấy _keyword_ từ _properties_ có lưu luôn ở dạng _set_ không để đỡ phải convert.